examples/chat-in-the-browser/index.ts

```diff
@@ -0,0 +1,131 @@
+      connEncryption: [Secio],
+      streamMuxer: [Mplex],
+      peerDiscovery: [Boostrap]
+    },
+    config: {
+      peerDiscovery: {
+        bootstrap: {
+          enabled: true,
+          list: [
+            '/dns4/ams-1.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLer265NRgSp2LA3dPaeykiS1J6DifTC88f5uVQKNAd',
+            '/dns4/lon-1.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLMeWqB7YGVLJN3pNLQpmmEk35v6wYtsMGLzSr5QBU3',
+            '/dns4/sfo-3.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLPppuBtQSGwKDZT2M73ULpjvfd3aZ6ha4oFGL1KrGM',
+            '/dns4/sgp-1.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLSafTMBsPKadTEgaXctDQVcqN88CNLHXMkTNwMKPnu',
+            '/dns4/nyc-1.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLueR4xBeUbY9WZ9xGUUxunbKWcrNFTDAadQJmocnWm',
+            '/dns4/nyc-2.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLV4Bbm51jM9C4gDYZQ9Cy3U6aXMJDAbzgu2fzaDs64'
+          ]
+        }
+      }
+    }
+  })
+
+  // Our protocol identifier
+  const protocol = '/chat'
+
+  // UI elements
+  const status = document.getElementById('status')!
+  const output = document.getElementById('output')!
+  const txtSend = document.getElementById('txt_send')! as HTMLInputElement
+  const btnSend = document.getElementById('btn_send')! as HTMLButtonElement
+
+  output.textContent = ''
+
+  function log (txt: string) {
+    console.info(txt)
+    output.textContent += `${txt.trim()}\n`
+  }
+
+  // Add the signaling server address, along with our PeerId to our multiaddrs list
+  // libp2p will automatically attempt to dial to the signaling server so that it can
+  // receive inbound connections from other peers
+  const webrtcAddr = '/ip4/0.0.0.0/tcp/9090/wss/p2p-webrtc-star'
+  libp2p.peerInfo.multiaddrs.add(multiaddr(webrtcAddr))
+
+  // Listen for new peers
+  libp2p.on('peer:discovery', (peerInfo: PeerInfo) => {
+    log(`Found peer ${peerInfo.id.toB58String()}`)
+  })
+
+  // Listen for new connections to peers
+  let remotePeer: PeerInfo | null
+
+  libp2p.on('peer:connect', (peerInfo: PeerInfo) => {
+    log(`Connected to ${peerInfo.id.toB58String()}`)
+    libp2p
+      .dialProtocol(peerInfo, [protocol])
+      .then(({ stream }: { stream: Stream }) => {
+        log('dialed a stream')
+        remotePeer = peerInfo
+        btnSend.disabled = false
+      })
+  })
+
+  // Listen for peers disconnecting
+  libp2p.on('peer:disconnect', (peerInfo: PeerInfo) => {
+    log(`Disconnected from ${peerInfo.id.toB58String()}`)
+  })
+
+  await libp2p.start()
+  status.innerText = 'libp2p started!'
+  log(`libp2p id is ${libp2p.peerInfo.id.toB58String()}`)
+
+  const handleChat: ProtocolHandler = async ({ connection, stream }) => {
+    log(`handle chat from ${connection?.remotePeer.toB58String()}`)
+    const handledStream = stream
+    pipe(handledStream, async function (source: AsyncGenerator<any, any, any>) {
+      for await (const msg of source) {
+        log(`Received message: ${msg}`)
+        pipe([], handledStream)
```

Comment:
This isn't causing issues now because you're only sending 1 message, but if you send multiples this will fail. The pipe needs to be outside of the for loop. If you want to ack individual messages you would likely use the `it-pushable` approach as well, or a [transform](https://github.com/libp2p/js-libp2p/blob/master/doc/STREAMING_ITERABLES.md#transform) to send the response for each message.

Reply:
Yes, that makes sense,I’ll update the example.